src/lua/alloc.c

```diff
@@ -0,0 +1,171 @@
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "lua/utils.h"
+
+#include <lauxlib.h> /* luaL_error */
+
+#define LUA_MEMORY_DEFAULT (2LL * 1024 * 1024 * 1024)
+
+/**
+ * The default allocator function which is wrapped into a new one
+ * with the Lua memory limit checker.
+ */
+static lua_Alloc orig_alloc;
+/**
+ * The maximum allowed LuaJIT memory limit.
+ */
+static size_t memory = LUA_MEMORY_DEFAULT;
+/**
+ * Currently allocated memory amount by LuaJIT.
+ */
+static size_t allocated;
+
+/**
+ * Lua custom memory allocation function. It extends the original
+ * one with a memory counter and a limit check.
+ */
+static void *
+luaT_alloc(void *ud, void *ptr, size_t osize, size_t nsize)
+{
+	size_t new_allocated = allocated + nsize - osize;
+	if (new_allocated > memory) {
+		diag_set(OutOfMemory, nsize, "luaT_alloc", "Lua objects");
+		return NULL;
+	}
+
+	void *result = orig_alloc(ud, ptr, osize, nsize);
+
+	if (result != NULL || nsize == 0)
+		allocated = new_allocated;
+	else
+		diag_set(OutOfMemory, nsize, "luaT_alloc", "Lua objects");
+
+	return result;
+}
+
+int
+luaT_setallocmemory(size_t amount)
+{
+	if (amount < allocated) {
+		diag_set(LuajitError, "Cannot limit the Lua memory with values "
+			 "less than the currently allocated amount");
+		return -1;
+	}
+	memory = amount;
+	return 0;
+}
+
+void
+luaT_initalloc(struct lua_State *L)
+{
+	assert(L != NULL);
+
+	allocated = luaL_getgctotal(L);
+
+	void *orig_ud;
+	orig_alloc = lua_getallocf(L, &orig_ud);
+
+	assert(orig_alloc != NULL);
+
+	lua_setallocf(L, luaT_alloc, orig_ud);
+}
+
+/**
+ * alloc.getmemory() -- get the allocator memory limit.
+ */
+static int
+lbox_alloc_getmemory(struct lua_State *L)
+{
+	lua_pushinteger(L, memory);
+	return 1;
+}
+
+/**
+ * alloc.setmemory() -- set the allocator memory limit.
+ * Returns old memory limit.
+ */
+static int
+lbox_alloc_setmemory(struct lua_State *L)
+{
+	if (lua_gettop(L) < 1)
+		luaL_error(L, "Usage: alloc.setmemory(amount)");
+
+	long long amount = lua_tointeger(L, 1);
```

Comment:
Should it be `size_t` too (unsigned)? Does `lua_tointeger` handles values beyond 2^31-1 correctly?

Reply:
I've tried to use `size_t` though it works quite strange. So I changed the value type to ssize_t.

You may see quite similar approach is mentioned in [one of the comments above](https://github.com/tarantool/tarantool/pull/10223#discussion_r1853557675).

Full diff of the method:
```diff
@@ -115,10 +90,12 @@ lbox_alloc_getmemory(struct lua_State *L)
 static int
 lbox_alloc_setmemory(struct lua_State *L)
 {
-       if (lua_gettop(L) < 1)
-               luaL_error(L, "Usage: alloc.setmemory(amount)");
+       if (lua_gettop(L) < 1) {
+               diag_set(IllegalParams, "Usage: alloc.setmemory(amount)");
+               return luaT_error(L);
+       }

-       long long amount = lua_tointeger(L, 1);
+       ssize_t amount = luaL_checkinteger(L, 1);
```

Reply:
> I've added the test on setting 8Gb memory limit

But, please, don't actually allocate so much memory in a test. We're testing on various hardware and sometimes these are quite small machines.

Reply:
It just tests it's possible to specify such limit, not actually allocate this much RAM.

Reply:
OK. I don't bother about values beyond 2^63-1.